src/image_cache/mod.rs

```diff
@@ -289,86 +297,62 @@ impl ImageCache {
-				requested_frame_id = 0;
-			} else {
-				requested_frame_id = self.current_frame_idx as isize;
-			}
-			self.current_file_idx = new_file_index;
-			let mut tmp_cache = BTreeMap::new();
-			mem::swap(&mut self.texture_cache, &mut tmp_cache);
-
-			// Delete all entries that are outside the range of files around the current file
-			// allowed by the capacity.
-			// Walk through our list of directory entries sorted by their distance from the current
-			// file and in each step remove an entry from the cache until we reach the desired cache
-			// size
-			let mut sorted_files: Vec<_> = tmp_cache.into_iter().enumerate().collect();
-			sorted_files.sort_unstable_by_key(|(index, _)| {
-				(*index as isize - self.current_file_idx as isize).abs()
-			});
-			let mut remaining_capacity = self.total_capacity;
-			sorted_files.retain(|cached| {
-				let (_path, texture) = &cached.1;
-				let mut all_frames_size = 0;
-				// TODO consider retaining individual frames.
-				for t in texture.frames.iter() {
-					// Thew new file has to fit in the cache after this operation
-					// which is why we multiply the estimated size by two
-					let dimensions = (t.texture.width(), t.texture.height());
-					all_frames_size += get_image_size_estimate(dimensions) as isize;
-				}
-				let shoudl_retain = remaining_capacity > all_frames_size + self.curr_est_size;
-				if shoudl_retain {
-					remaining_capacity -= all_frames_size;
-				}
-				shoudl_retain
-			});
-			tmp_cache = sorted_files.into_iter().map(|(_, entry)| entry).collect();
-			mem::swap(&mut self.texture_cache, &mut tmp_cache);
-			self.remaining_capacity = remaining_capacity;
 		}
 
+		let new_file_index = self.get_file_index(&parent, &target_file_name)?;
+
+		let requested_frame_id = match frame_id {
+			Some(frame_id) => frame_id,
+			None if self.current_file_idx != new_file_index => 0,
+			None => self.current_frame_idx as isize,
+		};
+
+		self.current_file_idx = new_file_index;
+		self.refresh_cache();
 		self.try_getting_requested_image(display, self.current_file_idx, requested_frame_id)
 	}
 
+	fn refresh_cache(&mut self) {
+		let cache = mem::take(&mut self.texture_cache);
+
+		// Delete all entries that are outside the range of files around the current file
+		// allowed by the capacity.
+		// Walk through our list of directory entries sorted by their distance from the current
+		// file and in each step remove an entry from the cache until we reach the desired cache
+		// size
+		let mut sorted_files: Vec<_> = cache.into_iter().enumerate().collect();
+		sorted_files.sort_unstable_by_key(|&(index, _)| {
+			(index as isize - self.current_file_idx as isize).abs()
+		});
+		self.remaining_capacity = self.total_capacity;
+		sorted_files.retain(|(_, (_, texture))| {
+			// TODO consider retaining individual frames.
+			let all_frames_size: isize = texture
+				.frames
+				.iter()
+				.map(|t| get_image_size_estimate(t.texture.width(), t.texture.height()))
+				.sum();
```

Comment:
Isn't `get_anim_size_estimate` supposed to be used for this?

Reply:
Yes, I'll update it.